apollo-router-core/src/query_planner/mod.rs

```diff
@@ -395,48 +356,108 @@ impl FetchNode {
                             .unwrap_or_default()
                     })
                     .collect::<Object>(),
-            );
+                paths: None,
+            })
+        }
+    }
 
-            // We already validated that the service exists during planning
-            let fetcher = service_registry.get(service_name).unwrap();
-
-            let (res, _tail) = fetcher
-                .stream(
-                    Request::builder()
-                        .query(operation.clone())
-                        .variables(Arc::clone(&variables))
-                        .build(),
-                )
-                .await
-                .into_future()
-                .instrument(query_span)
-                .await;
-
-            match res {
-                Some(response) if !response.is_primary() => {
-                    Err(FetchError::SubrequestUnexpectedPatchResponse {
-                        service: service_name.to_owned(),
-                    })
-                }
-                Some(Response {
-                    data, mut errors, ..
-                }) => {
-                    let mut response = response
-                        .lock()
-                        .instrument(tracing::trace_span!("response_lock_wait"))
-                        .await;
+    async fn fetch_node<'a>(
+        &'a self,
+        data: &'a Value,
+        current_dir: &'a Path,
+        request: &'a Arc<Request>,
+        service_registry: Arc<dyn ServiceRegistry>,
+        schema: &'a Arc<Schema>,
+    ) -> Result<Value, FetchError> {
+        let FetchNode {
+            operation,
+            service_name,
+            ..
+        } = self;
 
-                    let span = tracing::trace_span!("response_insert");
-                    let _guard = span.enter();
-                    response.append_errors(&mut errors);
-                    response.insert_data(current_dir, data)?;
+        let query_span = tracing::info_span!("subfetch", service = service_name.as_str());
 
-                    Ok(())
-                }
-                None => Err(FetchError::SubrequestNoResponse {
+        let Variables { variables, paths } =
+            self.make_variables(data, current_dir, request, schema)?;
+
+        // We already checked that the service exists during planning
+        let fetcher = service_registry.get(service_name).unwrap();
```

Comment:
:yarn: proofs inside the expect

```suggestion
        let fetcher = service_registry.get(service_name).expect("we already checked that the service exists during planning; qed");
```

Reply:
:+1: